src/main/java/in/projecteka/consentmanager/ConsentManagerConfiguration.java

```diff
@@ -53,10 +64,60 @@
     private static final String CM_DEAD_LETTER_EXCHANGE = "cm-dead-letter-exchange";
     private static final String CM_DEAD_LETTER_ROUTING_KEY = "cm-dead-letter";
 
+    @ConditionalOnProperty(value = "consentmanager.cacheMethod", havingValue = "guava", matchIfMissing = true)
+    @Bean({"accessToken"})
+    public CacheAdapter<String, String> createLoadingCacheAdapterForAccessToken() {
+        return new LoadingCacheAdapter(createSessionCache(5));
+    }
+
+    @ConditionalOnProperty(value = "consentmanager.cacheMethod", havingValue = "guava", matchIfMissing = true)
+    @Bean({"refreshToken"})
+    public CacheAdapter<String, String> createLoadingCacheAdapterForRefreshToken() {
+        return new LoadingCacheAdapter(createSessionCache(30));
+    }
+
+    public LoadingCache<String, String> createSessionCache(int duration) {
+        return CacheBuilder
+                .newBuilder()
+                .expireAfterWrite(duration, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, String>() {
+                    public String load(String key) {
+                        return "";
+                    }
+                });
+    }
+
+
+    @ConditionalOnProperty(value = "consentmanager.cacheMethod", havingValue = "redis")
+    @Bean({"accessToken"})
+    public CacheAdapter<String, String> createRedisCacheAdapter(RedisOptions redisOptions) {
+        RedisClient redisClient = getRedisClient(redisOptions);
+        return new RedisCacheAdapter(redisClient, 5);
+    }
+
+    @ConditionalOnProperty(value = "consentmanager.cacheMethod", havingValue = "redis")
+    @Bean({"refreshToken"})
+    public CacheAdapter<String, String> createDayRedisCacheAdapter(RedisOptions redisOptions) {
+        RedisClient redisClient = getRedisClient(redisOptions);
+        return new RedisCacheAdapter(redisClient, 30);
+    }
```

Comment:
I understood that guava cache does not support expiry per key, rather it supports only global level and you want to use guava for local development that is fine. 

But Redis support expiry per key, and you should use that, and expiry of accessToken and refresh token should be what you get in first sessions response. And key should CLIENT\_REGISTRY\_ACCESS\_TOKEN and CLIENT\_REGISTRY\_REFRESH\_TOKEN, since we will have similar kind of thing for CM\_INTERNAL\_SERVICES like USER\_SERVICE\_ACCESS\_TOKEN, etc.

What stops you to having @bean for **RedisClient** and reuse it everywhere?

Reply:
and you should keep the access token as long as refresh token is valid, because Keycloak would expect you to pass accessToken along with refreshToken in order to the accessToken if I'm not wrong.

Reply:
No need of the access token, Keycloak only needs refreshToken, ClientId and Client Secret

Reply:
I dont think you need to pass client secret, that defeats the whole purpose, if I pass client secret why would I need RefreshToken?https://stackoverflow.com/questions/51386337/refresh-access-token-via-refresh-token-in-keycloak

Reply:
Ok, will try it out, change it

Reply:
I tried but, It does need the ClientSecret as well

Reply:
https://www.oauth.com/oauth2-servers/access-tokens/refreshing-access-tokens/

Reply:
If I look at the link what I put where they don't have client\_secret. I think I got why do you have to send client\_secret also, I think it's because are our clients `confidential.` If you uncheck and try, you should be able to get token without passing `client_secret`. 

Having said that, since we need to pass client\_secret too, I think we can drop the new API to get the access\_token using refresh\_token, and make it simple, 

1.   AccessToken exists use it, otherwise call the existing token API with client\_id and secret.

Sorry for the effort to create the new API.

Reply:
you still need to address the redisclient bean and setting expiry per key in case of redis.